### ISSUE TITLE
Install instructions appear to be inadequate

### ISSUE THREAD

**Issue by omegonthesane** [openglad/openglad#13]
The installation instructions in INSTALL have not been updated since '02. I've tried for the last several days to get the Github version of Openglad's code to compile, without success, and it's got to the point where I'm starting to doubt that the problem in fact exists between keyboard and chair.

In particular, the Linux instructions (which I attempted to follow with Cygwin) fail to mention that you need to run autoconf, while attempting to use VisualC++ results in failure to actually follow links given in the makefile, and ultimately in claiming that a file that is right there in the same directory which has been referred to with #include "something" is not, in fact, where it actually is.

So, basically, the install instructions may need updating.

**Comment by silverhammermba** [comment]
I'm having a hard time building on Linux using premake4.4 beta 5. It fails at the linking stage:

```
$ premake4 --includedirs=/usr/include/SDL2 gmake && make
==== Building openglad (debug) ====
openglad.make:461: warning: overriding recipe for target 'obj/Debug/loader.o'
openglad.make:345: warning: ignoring old recipe for target 'obj/Debug/loader.o'
openglad.make:465: warning: overriding recipe for target 'obj/Debug/parser.o'
openglad.make:357: warning: ignoring old recipe for target 'obj/Debug/parser.o'
Linking openglad
g++ -o ./openglad obj/Debug/*.o -L. -lSDL2 -lSDL2_mixer -lpng
obj/Debug/loader.o: In function `yaml_parser_load':
/home/max/proj/openglad/openglad/src/external/libyaml/src/loader.c:68: multiple definition of `yaml_parser_load'
obj/Debug/loader.o:/home/max/proj/openglad/openglad/src/external/libyaml/src/loader.c:68: first defined here
obj/Debug/parser.o: In function `yaml_parser_parse':
/home/max/proj/openglad/openglad/src/external/libyaml/src/parser.c:171: multiple definition of `yaml_parser_parse'
obj/Debug/parser.o:/home/max/proj/openglad/openglad/src/external/libyaml/src/parser.c:171: first defined here
obj/Debug/button.o: In function `vbutton::vbutton(int, int, int, int, int, int, std::string const&, char, int)':
/home/max/proj/openglad/openglad/src/button.cpp:187: undefined reference to `loader::create_pixieN(char, char)'
obj/Debug/button.o: In function `vbutton::set_graphic(char)':
/home/max/proj/openglad/openglad/src/button.cpp:231: undefined reference to `loader::create_pixieN(char, char)'
obj/Debug/button.o: In function `toggle_effect(std::string const&, std::string const&)':
/home/max/proj/openglad/openglad/src/button.cpp:525: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/button.cpp:525: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/button.cpp:526: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/button.cpp:526: undefined reference to `cfg_store::apply_setting(std::string const&, std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/button.cpp:528: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/button.cpp:528: undefined reference to `cfg_store::apply_setting(std::string const&, std::string const&, std::string const&)'
obj/Debug/button.o: In function `toggle_rendering_engine()':
/home/max/proj/openglad/openglad/src/button.cpp:533: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/button.cpp:533: undefined reference to `cfg_store::get_setting(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/button.cpp:541: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/button.cpp:541: undefined reference to `cfg_store::apply_setting(std::string const&, std::string const&, std::string const&)'
obj/Debug/button.o: In function `vbutton::do_call(int, int)':
/home/max/proj/openglad/openglad/src/button.cpp:620: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/button.cpp:620: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/button.cpp:650: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/button.cpp:650: undefined reference to `cfg_store::load_settings()'
obj/Debug/glad.o: In function `main':
/home/max/proj/openglad/openglad/src/glad.cpp:77: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/glad.cpp:77: undefined reference to `cfg_store::load_settings()'
/home/max/proj/openglad/openglad/src/glad.cpp:78: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/glad.cpp:78: undefined reference to `cfg_store::save_settings()'
/home/max/proj/openglad/openglad/src/glad.cpp:79: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/glad.cpp:79: undefined reference to `cfg_store::commandline(int&, char**&)'
obj/Debug/guy.o: In function `guy::update_derived_stats(walker*)':
/home/max/proj/openglad/openglad/src/guy.cpp:505: undefined reference to `loader::set_derived_stats(walker*, char, char)'
obj/Debug/guy.o: In function `guy::create_walker(screen*)':
/home/max/proj/openglad/openglad/src/guy.cpp:586: undefined reference to `loader::create_walker(char, char, screen*, bool)'
obj/Debug/input.o: In function `handle_key_event(SDL_Event const&)':
/home/max/proj/openglad/openglad/src/input.cpp:368: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/input.cpp:368: undefined reference to `cfg_store::load_settings()'
obj/Debug/level_data.o: In function `LevelData::LevelData(int)':
/home/max/proj/openglad/openglad/src/level_data.cpp:302: undefined reference to `loader::loader()'
obj/Debug/level_data.o: In function `LevelData::add_ob(char, char, bool)':
/home/max/proj/openglad/openglad/src/level_data.cpp:374: undefined reference to `loader::create_walker(char, char, screen*, bool)'
obj/Debug/level_data.o: In function `LevelData::add_fx_ob(char, char)':
/home/max/proj/openglad/openglad/src/level_data.cpp:388: undefined reference to `loader::create_walker(char, char, screen*, bool)'
obj/Debug/level_data.o: In function `LevelData::add_weap_ob(char, char)':
/home/max/proj/openglad/openglad/src/level_data.cpp:400: undefined reference to `loader::create_walker(char, char, screen*, bool)'
obj/Debug/level_data.o: In function `LevelData::load()':
/home/max/proj/openglad/openglad/src/level_data.cpp:1294: undefined reference to `loader::loader()'
obj/Debug/level_editor.o: In function `LevelEditorData::activate_mode_button(SimpleButton*)':
/home/max/proj/openglad/openglad/src/level_editor.cpp:1392: undefined reference to `loader::set_walker(walker*, char, char)'
/home/max/proj/openglad/openglad/src/level_editor.cpp:1414: undefined reference to `loader::set_walker(walker*, char, char)'
obj/Debug/level_editor.o: In function `LevelEditorData::display_panel(screen*)':
/home/max/proj/openglad/openglad/src/level_editor.cpp:1679: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/level_editor.cpp:1679: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/level_editor.cpp:1947: undefined reference to `loader::set_walker(walker*, char, char)'
/home/max/proj/openglad/openglad/src/level_editor.cpp:1967: undefined reference to `loader::set_walker(walker*, char, char)'
/home/max/proj/openglad/openglad/src/level_editor.cpp:1990: undefined reference to `loader::set_walker(walker*, char, char)'
obj/Debug/living.o: In function `living::act()':
/home/max/proj/openglad/openglad/src/living.cpp:173: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/living.cpp:173: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
obj/Debug/picker.o: In function `show_guy(int, int, short, short)':
/home/max/proj/openglad/openglad/src/picker.cpp:2690: undefined reference to `loader::create_walker(char, char, screen*, bool)'
obj/Debug/picker.o: In function `draw_toggle_effect_button(button&, std::string const&, std::string const&)':
/home/max/proj/openglad/openglad/src/picker.cpp:3149: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/picker.cpp:3149: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
obj/Debug/picker.o: In function `main_options()':
/home/max/proj/openglad/openglad/src/picker.cpp:3194: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/picker.cpp:3194: undefined reference to `cfg_store::get_setting(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/picker.cpp:3230: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/picker.cpp:3230: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/picker.cpp:3231: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/picker.cpp:3231: undefined reference to `cfg_store::save_settings()'
obj/Debug/results_screen.o: In function `show_guy(int, guy*, short, short)':
/home/max/proj/openglad/openglad/src/results_screen.cpp:250: undefined reference to `loader::create_walker(char, char, screen*, bool)'
obj/Debug/screen.o: In function `screen::screen(short)':
/home/max/proj/openglad/openglad/src/screen.cpp:150: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/screen.cpp:150: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
obj/Debug/screen.o: In function `screen::set_walker(walker*, char, char)':
/home/max/proj/openglad/openglad/src/screen.cpp:1062: undefined reference to `loader::set_walker(walker*, char, char)'
obj/Debug/video.o: In function `video::video()':
/home/max/proj/openglad/openglad/src/video.cpp:42: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/video.cpp:42: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/video.cpp:47: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/video.cpp:47: undefined reference to `cfg_store::get_setting(std::string const&, std::string const&)'
obj/Debug/walker.o: In function `walker::fire()':
/home/max/proj/openglad/openglad/src/walker.cpp:766: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/walker.cpp:766: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
obj/Debug/walker.o: In function `draw_smallHealthBar(walker*, viewscreen*)':
/home/max/proj/openglad/openglad/src/walker.cpp:935: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/walker.cpp:935: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
obj/Debug/walker.o: In function `walker::do_heal_effects(walker*, walker*, short)':
/home/max/proj/openglad/openglad/src/walker.cpp:1775: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/walker.cpp:1775: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
obj/Debug/walker.o: In function `walker::do_hit_effects(walker*, walker*, short)':
/home/max/proj/openglad/openglad/src/walker.cpp:1785: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/walker.cpp:1785: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/walker.cpp:1796: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/walker.cpp:1796: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/walker.cpp:1826: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/walker.cpp:1826: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/walker.cpp:1829: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/walker.cpp:1829: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
obj/Debug/walker.o: In function `walker::special()':
/home/max/proj/openglad/openglad/src/walker.cpp:2584: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/walker.cpp:2584: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
/home/max/proj/openglad/openglad/src/walker.cpp:3889: undefined reference to `cfg'
/home/max/proj/openglad/openglad/src/walker.cpp:3889: undefined reference to `cfg_store::is_on(std::string const&, std::string const&)'
collect2: error: ld returned 1 exit status
openglad.make:242: recipe for target 'openglad' failed
make[1]: *** [openglad] Error 1
Makefile:16: recipe for target 'openglad' failed
make: *** [openglad] Error 2
```

I don't get how those references are missing, because I see `cfg` and `create_pixieN` defined in `parser.cpp` and `loader.cpp`, respectively.

**Comment by silverhammermba** [comment]
Apparently this issue (having source files in different directories with the same name) is [fixed](http://sourceforge.net/p/premake/feature-requests/73/) with premake 5. For now it appears that renaming openglad's `parser*` and `loader*` files allows it to link successfully.

**Comment by grimfang4** [comment]
Thanks for bringing this build issue to our attention!  This is certainly because both libyaml and Openglad have similarly-named files (loader.c, parser.c, loader.cpp, and parser.cpp) and your version of Premake seems to have a bug that doesn't create proper directory tree for object files.

Aren't buffer overflows THE GREATEST?!  :)
Any chance you could give more info on them, say, via valgrind output?

**Comment by silverhammermba** [comment]
I can work on getting some valgrind output. Would you be amenable to a pull request of file name changes so that there isn't a conflict with premake4?

**Comment by grimfang4** [comment]
Yeah, I think that would be reasonable.

**Comment by R-F-R** [comment]
Are there any updates on this issue? I am running into the same problem as silverhammermba here. I'm trying to get openglad to run on my RetroPie. I tried premake, but that stuff goes a little over my head to be quite honest.

**Comment by zardus** [comment]
Can you try opening the codeblocks file? That method worked great for me last time I tried it.

**Comment by R-F-R** [comment]
Ok, I really am trying my best to read the documentation and find my way around it, as I know it is annoying if people don't try to figure things out on their own. However, I know next to nothing about this, so you'll have to bear with me a little.
I did download CodeBlocks, got it running, downloaded the entire master folder and opened the "cb_Openglad.cbp" file. Then I found all the ".h" files he complained about not having when trying to build and dumped them in the "src" folder. 
Now I'm kind of at an impasse, because when I try to build, he has a whole host of complaints, none of which I understand. There is talk of "conflicting declarations" and "redeclarations" and the such. If I had to guess, I would say that stands in some relation with my horrible approach to "some configuration as to where SDL and so forth are".  Do you think you could shed some helping light on that?

> ||=== Build: Windows Debug SDL2 in Openglad (compiler: GNU GCC Compiler) ===|
D:\openglad-master\src\SDL_config_win32.h|71|warning: "SDL_HAS_64BIT_TYPE" redefined|
D:\openglad-master\src\SDL_types.h|61|note: this is the location of the previous definition|
D:\openglad-master\src\SDL_stdinc.h|79|warning: "SDL_TABLESIZE" redefined|
D:\openglad-master\src\SDL_types.h|34|note: this is the location of the previous definition|
D:\openglad-master\src\input.h|33|error: 'SDL_Keycode' was not declared in this scope|
D:\openglad-master\src\input.h|224|note: in expansion of macro 'SDLKey'|
D:\openglad-master\src\input.h|33|error: 'SDL_Keycode' has not been declared|
D:\openglad-master\src\input.h|225|note: in expansion of macro 'SDLKey'|
D:\openglad-master\src\SDL_stdinc.h|94|error: redeclaration of 'SDL_FALSE'|
D:\openglad-master\src\SDL_types.h|38|note: previous declaration 'SDL_bool SDL_FALSE'|
D:\openglad-master\src\SDL_stdinc.h|95|error: redeclaration of 'SDL_TRUE'|
D:\openglad-master\src\SDL_types.h|39|note: previous declaration 'SDL_bool SDL_TRUE'|
D:\openglad-master\src\SDL_stdinc.h|96|error: conflicting declaration 'typedef enum SDL_bool SDL_bool'|
D:\openglad-master\src\SDL_types.h|40|note: previous declaration as 'typedef enum SDL_bool SDL_bool'|
D:\openglad-master\src\SDL_stdinc.h|147|error: redeclaration of 'DUMMY_ENUM_VALUE'|
D:\openglad-master\src\SDL_types.h|110|note: previous declaration 'SDL_DUMMY_ENUM DUMMY_ENUM_VALUE'|
D:\openglad-master\src\SDL_stdinc.h|148|error: conflicting declaration 'typedef enum SDL_DUMMY_ENUM SDL_DUMMY_ENUM'|
D:\openglad-master\src\SDL_types.h|111|note: previous declaration as 'typedef enum SDL_DUMMY_ENUM SDL_DUMMY_ENUM'|
D:\openglad-master\src\button.h||In constructor 'button::button(Sint32, Sint32, Sint32, Sint32, Sint32, Sint32, const MenuNav&, bool, bool)':|
D:\openglad-master\src\input.h|35|error: 'SDL_SCANCODE_UNKNOWN' was not declared in this scope|
D:\openglad-master\src\button.h|95|note: in expansion of macro 'KEYSTATE_UNKNOWN'|
||=== Build failed: 8 error(s), 2 warning(s) (0 minute(s), 0 second(s)) ===|

Have I dun goofed?

**Comment by grimfang4** [comment]
Those errors make it seem like you have the wrong version of SDL (e.g. 1.2).  Did you get SDL 2.0?

I updated the build instructions in INSTALL: ceaa7407b9fbf9340199fd8beebb3f2bcf59a576
Hopefully that will help.

**Comment by R-F-R** [comment]
Thanks for all your work, and the update to the install file. You are right, I did use the wrong version of SDL, idiot that I am. I still get errors, though:

> ||=== Build: Windows Debug SDL2 in Openglad (compiler: GNU GCC Compiler) ===|
C:\Users\Dr-Bubbles\Desktop\opengladm\src\glad.cpp||In function 'short int new_score_panel(screen*, short int)':|
C:\Users\Dr-Bubbles\Desktop\opengladm\src\glad.cpp|610|error: 'strcpy' was not declared in this scope|
C:\Users\Dr-Bubbles\Desktop\opengladm\src\glad.cpp|611|error: 'strlen' was not declared in this scope|
C:\Users\Dr-Bubbles\Desktop\opengladm\src\glad.cpp|612|error: 'strcpy' was not declared in this scope|
C:\Users\Dr-Bubbles\Desktop\opengladm\src\glad.cpp|614|error: 'strcpy' was not declared in this scope|
C:\Users\Dr-Bubbles\Desktop\opengladm\src\glad.cpp|620|error: 'strcpy' was not declared in this scope|
C:\Users\Dr-Bubbles\Desktop\opengladm\src\glad.cpp|716|error: 'strcmp' was not declared in this scope|
||=== Build failed: 6 error(s), 0 warning(s) (0 minute(s), 0 second(s)) ===|

I also tried the premake method you described in the install file, but even though I did install all the SDL packages via apt-get, your suggested command doesn't work for me. Neither does the one where I have to manually include the folders where SDL is installed, but I think that failure is due to my lacking understanding of the linux file system... I'll report back on that front, as soon as I am reasonably sure I did everything to exclude user error ;)
Anyway, thx for your patience and help!

**Comment by grimfang4** [comment]
I tried to build and got the same errors.  I added more includes for those missing standard functions so that it builds successfully for me in a4d41c0cf99e33550c2501d8ec7669898ff67f5b.

**Comment by R-F-R** [comment]
Well, I'm getting further than I did before. I am running into all kinds of problems now with the PNG_EXPORT. I'm guessing I'm using the wrong version of libpng. Which one do you use? Do I have to build pnglibconf.h with any special settings? 
Thanks for all the help.

**Comment by grimfang4** [comment]
All libpng versions in the 1.6.x series should work.  I have libpng built separately as a dynamic library and link Openglad to it.  What are you trying to do with libpng?  What errors are you getting?

Since you bring it up, Openglad only uses libpng for saving screenshots.  Assuming we want to keep that functionality, it wouldn't take much to drop libpng and use stb-image-write instead to avoid the heavier dependency.

**Comment by R-F-R** [comment]
Well, I'm using 1.6.25, so that should work. I think the problem is between the screen and the chair here. Maybe it's better if I try to reinstall the libraries, I think I messed that up. I'll give that a try and let you know what happened.

**Comment by zardus** [comment]
Hopefully cmake (which we use now) is reasonable enough! Docs here: https://github.com/openglad/openglad/blob/master/docs/INSTALL.md
